File: frontend/src/types/greeks.ts
export interface GreeksSurfacePoint {
  underlyingPrice: number;
  delta: number;
  gamma: number;
  theta: number;
  vega: number;
  rho: number;
  theoreticalPrice: number;
}

export interface GreeksSurfaceResponse {
  points: GreeksSurfacePoint[];
  optionType: string;
  strike: number;
  dte: number;
  iv: number;
}

export interface TimeDecayPoint {
  dte: number;
  delta: number;
  gamma: number;
  theta: number;
  vega: number;
  theoreticalPrice: number;
}

export interface TimeDecayResponse {
  points: TimeDecayPoint[];
  optionType: string;
  strike: number;
  underlyingPrice: number;
  iv: number;
}

export interface IvImpactPoint {
  iv: number;
  delta: number;
  gamma: number;
  theta: number;
  vega: number;
  theoreticalPrice: number;
}

export interface IvImpactResponse {
  points: IvImpactPoint[];
  optionType: string;
  strike: number;
  dte: number;
  underlyingPrice: number;
}

export type GreekKey = "delta" | "gamma" | "theta" | "vega" | "rho";

File: backend/app/models/greeks.py
from __future__ import annotations

from pydantic import BaseModel

from app.models.option import CamelModel


class GreeksSurfaceRequest(BaseModel):
    option_type: str
    strike: float
    expiration: str
    iv: float
    underlying_price: float
    price_range: tuple[float, float] | None = None


class GreeksSurfacePoint(CamelModel):
    underlying_price: float
    delta: float
    gamma: float
    theta: float
    vega: float
    rho: float
    theoretical_price: float


class GreeksSurfaceResponse(CamelModel):
    points: list[GreeksSurfacePoint]
    option_type: str
    strike: float
    dte: int
    iv: float


class TimeDecayRequest(BaseModel):
    option_type: str
    strike: float
    dte: int
    iv: float
    underlying_price: float


class TimeDecayPoint(CamelModel):
    dte: int
    delta: float
    gamma: float
    theta: float
    vega: float
    theoretical_price: float


class TimeDecayResponse(CamelModel):
    points: list[TimeDecayPoint]
    option_type: str
    strike: float
    underlying_price: float
    iv: float


class IvImpactRequest(BaseModel):
    option_type: str
    strike: float
    dte: int
    underlying_price: float
    iv_range: tuple[float, float] | None = None


class IvImpactPoint(CamelModel):
    iv: float
    delta: float
    gamma: float
    theta: float
    vega: float
    theoretical_price: float


class IvImpactResponse(CamelModel):
    points: list[IvImpactPoint]
    option_type: str
    strike: float
    dte: int
    underlying_price: float
